packages/toast/src/ToastContainer/ToastContainer.tsx

```diff
@@ -0,0 +1,282 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { Transition, TransitionGroup } from 'react-transition-group';
+
+import { cx } from '@leafygreen-ui/emotion';
+import {
+  useBackdropClick,
+  useDynamicRefs,
+  useIdAllocator,
+  useMutationObserver,
+  useStateRef,
+} from '@leafygreen-ui/hooks';
+import { useDarkMode } from '@leafygreen-ui/leafygreen-provider';
+import {
+  createSyntheticEvent,
+  createUniqueClassName,
+} from '@leafygreen-ui/lib';
+import Portal from '@leafygreen-ui/portal';
+import { transitionDuration } from '@leafygreen-ui/tokens';
+
+import { TOAST } from '../constants';
+import { InternalToast } from '../InternalToast';
+import { ToastId, ToastStack, useToast } from '../ToastContext';
+
+import {
+  NotificationBar,
+  notificationBarTransitionStyles,
+} from './NotificationBar';
+import {
+  getContainerStatefulStyles,
+  getToastHoverStyles,
+  getToastTransitionStyles,
+  getToastUnhoveredStyles,
+  scrollContainerExpandedStyles,
+  scrollContainerStyles,
+  toastContainerStyles,
+  toastContainerVisibleStyles,
+} from './ToastContainer.styles';
+import {
+  getDividedStack,
+  useToastHeights,
+  useToastTimers,
+  useToastTransitions,
+} from './utils';
+
+export const toastPortalClassName = createUniqueClassName('toast-portal');
+
+/**
+ * ToastContainer is responsible for rendering the stack of toasts provided
+ */
+export const ToastContainer = ({ stack }: { stack: ToastStack }) => {
+  const { popToast, getToast } = useToast();
+  const regionId = useIdAllocator({ id: 'lg-toast-region' });
+  const stackSize = stack.size;
+  const toastContainerRef = useRef<HTMLDivElement>(null);
+  const scrollContainerRef = useRef<HTMLDivElement>(null);
+  const getToastRef = useDynamicRefs<HTMLDivElement>({ prefix: 'toast' });
+  const { theme } = useDarkMode();
+  const [isHovered, setHoveredState] = useState(false);
+  const setHovered = () => setHoveredState(true);
+  const setUnhovered = () => setHoveredState(false);
```

Comment:
Are two state variables needed here? Could you have one state typed with `boolean | undefined` instead?

Reply:
My fault! I misread. I thought there were two different state variables in use here.